### PR TITLE
Tooltips

### PR DESCRIPTION
**Pull request by larsiusprime**
Add support for tooltips, <inject>, and <default>

### CODE REVIEW

**Comment by larsiusprime** [comment]
@Gama11 Nope, that'd have saved me some typing!

**Comment by MSGhero** [comment]
Closes #102

**Comment by larsiusprime** [comment]
Incidentally, can we remove the Haxe 3.1.3 checks from Travis? Do we still need them? That's where all our false positive Travis fails come from.

**Comment by Gama11** [comment]
They're not exactly "false positives" - we test Haxe 3.1.3 since it's technically still supported (there hasn't really been any reason to drop the support for it, only a few very minor changes were needed for compatiblity).

**Comment by larsiusprime** [comment]
Hmmm. So I should go in and fix anything that causes it to fail on 3.1.3?

**Comment by larsiusprime** [comment]
I think this is what it's complaining about:

https://github.com/HaxeFlixel/flixel-ui/blob/lars_latest/flixel/addons/ui/FlxUI.hx#L580

How to fix this in Haxe 3.1.3?

**Comment by Gama11** [comment]
I'm not sure... Only being compatible with Haxe 3.2.0 wouldn't be a big deal since we actually have a heplful error message for that in the FlxDefines macro (if you were to try and compile with a Haxe version that's too old, compilation fails with "The minimum required Haxe version for HaxeFlixel is [version]. Please install a newer version." (We didn't have this in the past, which led to some cryptic error messages and a lot of confusion on the user end).

There are a couple of advantages I can see to dropping Haxe 3.1.3 support:
- shorter travis build times
- maintenance potentially becomes a little easier (see this issue)
- there are a few conditionals in the core that we could remove (barely noteworthy though).

Not sure what the reasons for wanting to stay on version 3.1.3 would be.. Some newly introduced Haxe bug maybe? But in general 3.2.0 should be more stable (the mentioned conditionals are a workaround for a haxe bug, for example).

**Comment by larsiusprime** [comment]
I bet if we just drop support we'll get at least a few users suddenly wondering what's wrong, but... I suppose "dropped support" just means the new, bleeding-edge dev version doesn't support it, which isn't the same as retroactively pushing it out to existing users and breaking stuff?

IF we do keep 3.1.3 support I can just wall off the new "inject" feature to be 3.2.0+ only.

**Comment by Gama11** [comment]
Well, the error message should tell them "what's wrong"? Or do you mean something else?

**Comment by larsiusprime** [comment]
Yeah I just mean, it's not like people tracking the current dev branch aren't used to keeping up with the latest stuff right? I think it's okay to drop 3.1.3 unless you have strong reason to do otherwise.

File: flixel/addons/ui/FlxUITooltip.hx [hunk 1]
```diff
@@ -0,0 +1,570 @@
...
+			autoSizeHorizontal: s.autoSizeHorizontal,
+			leftPadding  : s.leftPadding,
+			rightPadding : s.rightPadding,
+			topPadding   : s.topPadding,
+			bottomPadding: s.bottomPadding
+		}
+	}
+}
+
+typedef ToolTipStyle = {
```
Comment (Gama11) [inline flixel/addons/ui/FlxUITooltip.hx]: Any reason for the `@:optional var` syntax here? Pretty sure it's equivalent to having a `?` in front like you did in `ToolTipData `, but much shorter.

Comment (larsiusprime) [inline flixel/addons/ui/FlxUITooltip.hx]: I put it there b/c I forgot the syntax and that's what the api said, is ? strictly equivalent? If so I'll use that.

Comment (Gama11) [inline flixel/addons/ui/FlxUITooltip.hx]: Which API?

**Comment by Gama11** [comment]
I suppose we could have a minimum required version (3.1.3, compiler error) and a minimum recommended version (3.2.0, compiler warning)? The offending xml code would still need conditionals in that case though...

File: flixel/addons/ui/FlxUITooltip.hx (hunk 1 above)
Comment (larsiusprime) [inline flixel/addons/ui/FlxUITooltip.hx]: The Haxe API docs.

Comment (larsiusprime) [inline flixel/addons/ui/FlxUITooltip.hx]: http://haxe.org/manual/type-system-typedef.html

**Comment by MSGhero** [comment]
3.2 did introduce some bugs, but I doubt anyone on dev branch would mind.

File: flixel/addons/ui/FlxUITooltip.hx (hunk 1 above)
Comment (larsiusprime) [inline flixel/addons/ui/FlxUITooltip.hx]: Fixed.

**Comment by Gama11** [comment]
Hm, I can't even compile with Haxe 3.1.3 right now... Some error in `ArrayBufferView` about missing fields:

```
C:/HaxeToolkit/haxe/lib/lime/2,4,9/lime/utils/ArrayBufferView.hx:609: characters 12-27 : haxe.io.Bytes has no field setFloat
```

Must be something on my end, since it compiles on travis..

**Comment by larsiusprime** [comment]
I've hit up just about everything I can find with this now. Any further comments?

**Comment by larsiusprime** [comment]
(note that if travis fails, it's probably b/c I'm waiting on this PR to flixel: https://github.com/HaxeFlixel/flixel/pull/1565)

**Comment by larsiusprime** [comment]
Since this is flixel-ui and it's passing all the 3.2.0 checks, I think I'm just gonna merge this. Let me know if there's any trouble.

**Pull request merged**